sql/analyzer/join_search.go

```diff
@@ -0,0 +1,495 @@
+func colsIncludeTable(cols []*expression.GetField, table string) bool {
+	for _, col := range cols {
+		if strings.ToLower(col.Table()) == table {
+			return true
+		}
+	}
+	return false
+}
+
+// Generates all permutations of the slice given.
+func permutations(ints []int) [][]int{
+	var helper func([]int, int)
+	var res [][]int
+
+	helper = func(arr []int, n int){
+		if n == 1{
+			tmp := make([]int, len(arr))
+			copy(tmp, arr)
+			res = append(res, tmp)
+		} else {
+			for i := 0; i < n; i++{
+				helper(arr, n - 1)
+				if n % 2 == 1{
+					tmp := arr[i]
+					arr[i] = arr[n - 1]
+					arr[n - 1] = tmp
+				} else {
+					tmp := arr[0]
+					arr[0] = arr[n - 1]
+					arr[n - 1] = tmp
+				}
+			}
+		}
+	}
+	helper(ints, len(ints))
+	return res
+}
+
+// joinSearchParams is a simple struct to track available tables and join conditions during a join search
+type joinSearchParams struct {
+	tables               []string
+	usedTableIndexes     []int
+	joinConds            []*joinCond
+	usedJoinCondsIndexes []int
+}
+
+func (js *joinSearchParams) copy() *joinSearchParams {
+	usedTableIndexesCopy := make([]int, len(js.usedTableIndexes))
+	copy(usedTableIndexesCopy, js.usedTableIndexes)
+	usedJoinCondIndexesCopy := make([]int, len(js.usedJoinCondsIndexes))
+	copy(usedJoinCondIndexesCopy, js.usedJoinCondsIndexes)
+	return &joinSearchParams{
+		tables:               js.tables,
+		usedTableIndexes:     usedTableIndexesCopy,
+		joinConds:            js.joinConds,
+		usedJoinCondsIndexes: usedJoinCondIndexesCopy,
+	}
+}
+
+func (js *joinSearchParams) tableIndexUsed(i int) bool {
+	return indexOfInt(i, js.usedTableIndexes) >= 0
+}
+
+func (js *joinSearchParams) joinCondIndexUsed(i int) bool {
+	return indexOfInt(i, js.usedJoinCondsIndexes) >= 0
+}
+
+// A joinSearchNode is a simplified type representing a join tree node, which is either an internal node (a join) or a
+// leaf node (a table). The top level node in a join tree is always an internal node. Every internal node has both a
+// left and a right child.
+type joinSearchNode struct {
+	table    string            // empty if this is an internal node
+	joinCond *joinCond    // nil if this is a leaf node
+	parent   *joinSearchNode   // nil if this is the root node
+	left     *joinSearchNode   // nil if this is a leaf node
+	right    *joinSearchNode   // nil if this is a leaf node
+	params   *joinSearchParams // search params that assembled this node
+}
+
+// used to mark the left or right branch of a node as being targeted for assignment
+var childTargetNode = &joinSearchNode{}
+
+// tableOrder returns the order of the tables in this part of the tree, using an in-order traversal
+func (n *joinSearchNode) tableOrder() []string {
+	if n == nil {
+		return nil
+	}
+
+	if len(n.table) > 0 {
+		return []string{n.table}
+	}
+
+	var tables []string
+	tables = append(tables, n.left.tableOrder()...)
+	tables = append(tables, n.right.tableOrder()...)
+	return tables
+}
+
```

Comment:
Nit: does

```go
func (n *joinSearchNode) isLeaf() bool {
    return len(n.table) > 0
}
```

make the following functions a little bit easier to read?

Reply:
Great idea, thx